src/sql/workbench/browser/modal/modal.ts

```diff
@@ -168,40 +198,54 @@ export abstract class Modal extends Disposable implements IThemable {
 
 	/**
 	 * Build and render the modal, will call {@link Modal#renderBody}
+	 *
 	 */
 	public render() {
 		let builderClass = '.modal.fade';
-		if (this._modalOptions.isFlyout) {
-			builderClass += '.flyout-dialog';
-		}
+		builderClass += this._modalOptions.dialogStyle === 'flyout' ? '.flyout-dialog'
+			: this._modalOptions.dialogStyle === 'callout' ? '.callout-dialog'
+				: this._modalOptions.dialogStyle === 'calloutCompact' ? '.callout-dialog.compact'
+					: '';
 
 		this._bodyContainer = DOM.$(`${builderClass}`, { role: 'dialog', 'aria-label': this._title });
 		const top = this.layoutService.offset?.top ?? 0;
 		this._bodyContainer.style.top = `${top}px`;
 		this._modalDialog = DOM.append(this._bodyContainer, DOM.$('.modal-dialog'));
-		this._modalContent = DOM.append(this._modalDialog, DOM.$('.modal-content'));
+
+		if (this._modalOptions.dialogStyle === 'callout' || this._modalOptions.dialogStyle === 'calloutCompact') {
+			let arrowClass = `.arrow.from-${this._modalOptions.dialogPosition}`;
+			this._modalContent = DOM.append(this._modalDialog, DOM.$(`.modal-content${arrowClass}`));
+		} else {
+			this._modalContent = DOM.append(this._modalDialog, DOM.$('.modal-content'));
+		}
 
 		if (typeof this._modalOptions.width === 'number') {
 			this._modalDialog.style.width = `${this._modalOptions.width}px`;
 		} else {
 			this._modalDialog.classList.add(`${this._modalOptions.width}-dialog`);
 		}
 
+		if (this._modalOptions.dialogStyle === 'callout' || this._modalOptions.dialogStyle === 'calloutCompact') {
```

Comment:
For the future - in general when you have different functionality like this it's much better to keep all specific logic like this in the superclass. Modal shouldn't have to know anything about the implementation of a callout dialog for example.